TPP.Core/Moderation/Moderator.cs

```diff
@@ -0,0 +1,172 @@
+            {
+                return new RuleResult.DeleteMessage();
+            }
+
+            return new RuleResult.Nothing();
+        }
+
+        public async Task<bool> Check(Message message)
+        {
+            bool deleteMessage = false;
+            (RuleResult.Timeout, IModerationRule)? timeoutAndRule = null;
+
+            List<(RuleResult, IModerationRule)> pointResults = new();
+
+            void ProcessResult(RuleResult result, IModerationRule rule)
+            {
+                if (result is RuleResult.GivePoints givePoints)
+                    pointResults.Add((ApplyPoints(message.User, givePoints.Points, givePoints.Reason), rule));
+                else if (result is RuleResult.DeleteMessage)
+                    deleteMessage = true;
+                else if (result is RuleResult.Timeout resultTimeout)
+                    timeoutAndRule = (resultTimeout, rule);
+                else if (result is RuleResult.Nothing) { }
+                else
+                    _logger.LogWarning($"unhandled moderator rule result type '{result.GetType()}'");
+            }
+
+            foreach (IModerationRule? rule in _rules)
+            {
+                RuleResult result = rule.Check(message);
+                ProcessResult(result, rule);
+            }
+            while (pointResults.Any())
+            {
+                (RuleResult result, IModerationRule rule) = pointResults.First();
+                pointResults.RemoveAt(0);
+                ProcessResult(result, rule);
+            }
+
+            if (timeoutAndRule.HasValue)
+            {
+                (RuleResult.Timeout timeout, IModerationRule rule) = timeoutAndRule.Value;
+                Duration timeoutDuration = await CalculateTimeoutDuration(message.User);
+                await _executor.Timeout(message.User, timeout.Message, timeoutDuration);
+                await _modLogRepo.LogModAction(message.User, timeout.Message, rule.Id, _clock.GetCurrentInstant());
+                return false;
+            }
+            else if (deleteMessage)
+            {
+                if (message.Details.MessageId != null)
+                    await _executor.DeleteMessage(message.Details.MessageId);
+                else
+                    // Regular messages should always have an id. Whispers don't, but shouldn't be checked by modbot.
+                    _logger.LogWarning($"Modbot cannot delete message because it's missing a message id: {message}");
+                return false;
+            }
+            return true;
+        }
+
+        private async Task<Duration> CalculateTimeoutDuration(User user)
+        {
+            Instant cutoff = _clock.GetCurrentInstant() - RecentTimeoutsLimit;
+            long recentBans = await _modLogRepo.CountRecentBans(user, cutoff);
+
+            Duration duration = InitialTimeoutDuration;
+            long increases = Math.Max(0, recentBans - _freeTimeouts);
+            duration *= increases + 1;
```

Comment:
I thought so too, but it has been this way since 2016-09-27, introduced by a change from streamer.
I'm not sure if that was deliberate or accidental. It used to basically be a loop over `increases` where it did `duration *= 2` for each iteration

Reply:
OK, I think we should ask what we want now then. I'll bring it up in #Policy

Reply:
OK, #policy doesn't care, so I'm going to say go for the 2 minutes and move to doubling if modbot is seen as toothless